nixos/modules/system/boot/loader/bootloaders.nix

```diff
@@ -0,0 +1,66 @@
+{ config, lib, pkgs, ... }:
+with lib;
+
+let
+  cfg = config.boot.loader;
+  foldEnabledBootloaders = fn: start:
+    foldAttrs fn start
+    # Reverse order is a conscious choice to make it so that "last append" is first considered here.
+    # by default.
+      (map (moduleId: cfg.builders.${moduleId}) (reverseList cfg.enabled));
+in
+{
+  # Interface
+  options = {
+    boot.loader.builders = mkOption {
+      type = types.attrsOf (types.submodule ({ ... }: {
+        options = {
+          id = mkOption {
+            type = types.str;
+            description = lib.mdDoc "A string describing the bootloader name";
+          };
+          script = mkOption {
+            type = types.package;
+            description = lib.mdDoc "An executable script that ...";
+          };
+        };
+      }));
+      default = {};
+      example = literalExpression ''{ systemd-boot = finalSystemBootBuilder; }'';
+      internal = true;
+      description = lib.mdDoc ''
+        An attribute set of builder for any potential bootloader.
+        This is set internally by each bootloader with their
+        final system boot builder script.
+      '';
+    };
+    boot.loader.enabled = mkOption {
+      type = types.enum [
+        "systemd-boot"
+        "grub"
+        "generic-extlinux-compatible"
+      ];
```

Comment:
I don't know what the evaluation performance implications are, but I have a scheme in use for Mobile NixOS that makes "subnamespaces" self-describing with enums.

 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types.nix#L34
 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types.nix#L26-L28

And e.g.

 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types/depthcharge/default.nix#L55

Such a scheme would be *especially* useful to allow external bootloader interfaces to be added without modifying NixOS proper.

Reply:
I think it would be helpful to have @roberth from the getgo (and FWIW, if it was not clear, this PR has no intention to break things and backward compatibility yet, I am not even expecting to land it in 23.11 FWIW.)

Reply:
OK, I see better for the enums, interesting.

Reply:
It might even be possible without an intermediary option, because of the way `boot.loader.builders` is made:

```
default = type.enum builtins.attrNames boot.loader.builders
```

Reply:
This will break doc evaluation

Reply:
Is the point to pick a single bootloader?
In that case I'd do something like samueldr's suggestion, but using option type merging instead.

`loader.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum [];
  description = ''
    Which bootloader to install.
  '';
}
```

`grub.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum ["grub"];
}
```

`systemd-boot.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum ["systemd-boot"];
}
```

When all three modules are loaded, the types are merged such that the enum is allows the union of all values.

> This will break doc evaluation

I wouldn't be sure of that. You can have some dependencies on `config` in your types and descriptions etc, unless they depend on things that aren't available during doc evaluation, such as `pkgs` in NixOS iirc. You do have to be careful not to include default values in a way that suggests they're unchangeable, but that's not a problem here I think.

Reply:
This would definitely support multiple bootloaders, we went for plain subtrees as you can see in the latest, I'm not sure we need to use type merging at all in fact, what do you think of the latest?

> I wouldn't be sure of that. You can have some dependencies on config in your types and descriptions etc, unless they depend on things that aren't available during doc evaluation, such as pkgs in NixOS iirc. You do have to be careful not to include default values in a way that suggests they're unchangeable, but that's not a problem here I think.

Oh, interesting. In this situation, we let the "subtree" express the availabilities of "loader types".

> Is the point to pick a single bootloader?

We want to offer any combination of multiple bootloaders.

Reply:
Currently, I'm not using option type merging but I will keep this option in mind and feel free to propose it on the current iteration of the code.